File: src/main/java/minic/ui/MiniCHoverInspectorContent.java
package minic.ui;

import minic.uiapi.UiSourceSpanDto;

import java.util.List;
import java.util.Objects;

/**
 * 底部 hover inspector 的一次展示内容。
 *
 * @param title 标题
 * @param metadata 元数据行
 * @param source 源码全文
 * @param range 当前元素对应的源码范围；可为 {@code null}
 * @param explanation 右侧解释文本
 */
public record MiniCHoverInspectorContent(
        String title,
        List<String> metadata,
        String source,
        UiSourceSpanDto range,
        String explanation
) {
    public MiniCHoverInspectorContent {
        Objects.requireNonNull(title, "title");
        Objects.requireNonNull(metadata, "metadata");
        Objects.requireNonNull(source, "source");
        Objects.requireNonNull(explanation, "explanation");
        metadata = List.copyOf(metadata);
    }

    /**
     * 空内容。
     */
    public static MiniCHoverInspectorContent empty() {
        return new MiniCHoverInspectorContent("", List.of(), "", null, "");
    }

    /**
     * 是否没有可显示内容。
     *
     * @return 空则为 {@code true}
     */
    public boolean emptyContent() {
        return title.isBlank() && metadata.isEmpty() && source.isBlank() && explanation.isBlank();
    }
}

File: uiweb/src/panel/MiniCHoverInspectorContent.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import type { UiSourceSpanDto } from "../translation/uiapi";

export const miniCHoverInspectorContentMirror = {
  "javaPath": "src/main/java/minic/uilocal/panel/MiniCHoverInspectorContent.java",
  "webPath": "uiweb/src/panel/MiniCHoverInspectorContent.ts",
  "packageName": "minic.uilocal",
  "exportName": "MiniCHoverInspectorContent",
  "kind": "record",
  "imports": [
    "java.util.List",
    "java.util.Objects",
    "minic.uiapi.UiSourceSpanDto"
  ],
  "fields": [],
  "methods": [
    {
      "name": "empty",
      "signature": "empty()"
    },
    {
      "name": "emptyContent",
      "signature": "emptyContent()"
    }
  ]
} as const satisfies JavaMirrorFile;

export class MiniCHoverInspectorContent {
  static readonly mirror = miniCHoverInspectorContentMirror;

  readonly mirror = miniCHoverInspectorContentMirror;

  readonly title: string;

  readonly metadata: readonly string[];

  readonly source: string;

  readonly range: UiSourceSpanDto | null;

  readonly explanation: string;

  constructor(
    title: string,
    metadata: readonly string[],
    source: string,
    range: UiSourceSpanDto | null,
    explanation: string,
  ) {
    this.title = title;
    this.metadata = [...metadata];
    this.source = source;
    this.range = range;
    this.explanation = explanation;
  }

  static empty(): MiniCHoverInspectorContent {
    return new MiniCHoverInspectorContent("", [], "", null, "");
  }

  emptyContent(): boolean {
    return (
      this.title.trim() === "" &&
      this.metadata.length === 0 &&
      this.source.trim() === "" &&
      this.explanation.trim() === ""
    );
  }

  summary(): string {
    return `MiniCHoverInspectorContent: ${this.mirror.methods.length} methods, ${this.mirror.fields.length} fields`;
  }
}

export default MiniCHoverInspectorContent;
